scipy/stats/_pagel.py

```diff
@@ -0,0 +1,492 @@
+
+    if ranks.ndim != 2:  # TODO: relax this to accept 3d arrays?
+        raise ValueError(f"`data` must be a 2d array.")
+    m, n = ranks.shape
+    if m < 2 or n < 3:
+        raise ValueError("Page's L is only appropriate for data with two "
+                        "or more rows and three or more columns.")
+
+    # generate predicted ranks if not provided, ensure valid NumPy array
+    if predicted_ranks is None:
+        predicted_ranks = np.arange(n, 0, -1)
+    else:
+        predicted_ranks = np.array(predicted_ranks, copy=False)
+        if (predicted_ranks.ndim < 1 or
+                (set(predicted_ranks) != set(range(1, n+1)) or
+                len(predicted_ranks) != n)):
+            raise ValueError(f"`predicted_ranks` must include each integer "
+                            f"from 1 to {n} (the number of columns in data) "
+                            f"exactly once.")
+
+    if type(ranked) is not bool:
+        raise TypeError("`ranked` must be boolean.")
+
+    # Calculate the L statistic
+    L = _l_vectorized(ranks, predicted_ranks)
+
+    # Calculate the p-value
+    if method == "auto":
+        method = _choose_method(ranks)
+    p_fun = methods[method]  # get the function corresponding with the method
+    p = p_fun(L, m, n)
+
+    pagel_result = scipy.stats.StatsTestResult()
+    pagel_result['statistic'] = L
+    pagel_result['pvalue'] = p
+    pagel_result['method'] = method
+    return pagel_result
+
+
+def _choose_method(ranks):
+    '''Choose method for computing p-value automatically'''
+    m, n = ranks.shape
+    if n > 8 or (m > 12 and n > 3) or m > 20:  # as in [1], [4]
+        method = "asymptotic"
+    else:
+        method = "exact"
+    return method
+
+
+def _l_vectorized(ranks, predicted_ranks):
+    '''Calculate's Page's L statistic for each page of a 3d array'''
+    colsums = ranks.sum(axis=-2, keepdims=True)
+    products = predicted_ranks * colsums
+    Ls = products.sum(axis=-1)
+    Ls = Ls[0] if Ls.size == 1 else Ls.ravel()
+    return Ls
+
+
+def _l_p_asymptotic(L, m, n):
+    '''Calculate the p-value of Page's L from the asymptotic distribution'''
+    # Using [1] as a reference, the asymptotic p-value would be calculated as:
+    # chi_L = (12*L - 3*m*n*(n+1)**2)**2/(m*n**2*(n**2-1)*(n+1))
+    # p = chi2.sf(chi_L, df=1, loc=0, scale=1)/2
+    # but this is insentive to the direction of the hypothesized ranking
+
+    # See [2] page 151
+    E0 = (m*n*(n+1)**2)/4
+    V0 = (m*n**2*(n+1)*(n**2-1))/144
+    Lambda = (L-E0)/np.sqrt(V0)
+    p = norm.sf(Lambda)
+    return p
+
+
+def _l_p_exact(L, m, n):
+    L, n, k = int(L), int(m), int(n)  # different papers use different symbols
+    a, b = (k*(k+1)*(k+2))//6, (k*(k+1)*(2*k+1))//6
+
+#    dir_path = os.path.dirname(os.path.realpath(__file__))
+    dir_path = os.path.abspath(os.path.dirname(__file__))
+    datafile = os.path.join(dir_path, "data_folder", "pagel_exact.npy")
+    all_pmfs = np.load(datafile, allow_pickle=True).item()
```

Comment:
Avoid loading every use; set `all_pmfs = None` in module, and if it's `None` load it here.